src/Language/Haskell/Liquid/Bare/DataType.hs

```diff
@@ -156,3 +158,34 @@ makeTyConEmbeds' :: TCEmb (Located Symbol) -> BareM (TCEmb TyCon)
 makeTyConEmbeds' z = M.fromList <$> mapM tx (M.toList z)
   where
     tx (c, y) = (, y) <$> lookupGhcTyCon c
+
+makeRecordSelectorSigs :: [(DataCon, Located DataConP)] -> BareM [(Var, Located SpecType)]
+makeRecordSelectorSigs dcs = concat <$> mapM makeOne dcs
+  where
+  makeOne (dc, Loc l l' dcp)
+    | null (dataConFieldLabels dc)
+    = return []
+    | otherwise = do
+        fs <- mapM lookupGhcVar (dataConFieldLabels dc)
+        return (fs `zip` ts)
+    where
+    ts   = [ Loc l l' (mkArrow (freeTyVars dcp) [] (freeLabels dcp)
+                               [(z, res, mempty)]
+                               (dropPreds (subst su t `strengthen` mt)))
+           | (x, t) <- reverse args -- NOTE: the reverse here is correct
+           , not (isFunTy t) -- NOTE: we only have measures for non-function fields
+           , let vv = rTypeValueVar t
+             -- the measure singleton refinement, eg `v = getBar foo`
+           , let mt = uReft (vv, PAtom Eq (EVar vv) (EApp (EVar x) (EVar z)))
+           ]
+
+    su   = mkSubst $ [ (x, EApp (EVar x) (EVar z)) | x <- xs ]
+    args = tyArgs dcp
+    xs   = map fst args
+    z    = "lq$recSel"
+    res  = dropPreds (tyRes dcp)
+
+    -- FIXME: this is clearly imprecise, but the preds in the DataConP seem
+    -- to be malformed. If we leave them in, tests/pos/kmp.hs fails with
+    -- a malformed predicate application. Niki, help!!
+    dropPreds = fmap (\(MkUReft r _ps ss) -> MkUReft r mempty ss)
```

Comment:
This is the error LH produces.
```
 tests/pos/kmp.hs:64:10-13: Error: Bad Type Specification
 assumed type KMP.alen :: forall a <rng :: Int-> a -> Prop>.
                          lq$recSel:(Arr <{VV : a<rng x1> | true}> a) -> {v : Int | v >= 0
                                                                                    && v == alen lq$recSel}
     Missing arguments on  rng
```
(Btw this looks more like an unbound variable rather than a "missing argument"..)

But if I disable the "missing argument" check, we get these final types.

```
KMP.A : forall a <rng :: GHC.Types.Int-> a -> Prop>.
         alen##KMP.A##r10i:{VV : GHC.Types.Int | VV >= 0} -> aval##KMP.A##r10i:(i:{VV : GHC.Types.Int | VV >= 0
                                                                                                        && VV < alen##KMP.A##r10i} -> {VV : a<rng i> | true}) -> {VV : (KMP.Arr <\lq_tmp$db##6 VV -> {VV : _<rng> | true}> a) | alen VV == alen##KMP.A##r10i}
```

```
 KMP.alen : forall a <rng :: GHC.Types.Int-> a -> Prop>.
            lq$recSel:(KMP.Arr <{VV : a<rng> | true}> a) -> {v##16 : GHC.Types.Int | v##16 >= 0
                                                                                     && v##16 == alen lq$recSel}
```

Ah, it turns out the `rng` application to `Arr` is not quite the same. Curious, when you construct the datacon type do you do any special processing of abstract predicates? I was just extracting the types directly from the DataConP.